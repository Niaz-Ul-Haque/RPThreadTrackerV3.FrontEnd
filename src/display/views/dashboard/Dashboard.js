import React, { Component } from 'react';
import {
	Row, Col
} from 'reactstrap';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import AtAGlanceCard from './components/at-a-glance/AtAGlanceCard';
import RecentActivityCard from './components/recent-activity/RecentActivityCard';
import YourCharactersCard from './components/your-characters/YourCharactersCard';
import TrackerSupportCard from './components/tracker-support/TrackerSupportCard';
import RandomThreadCard from './components/random-thread/RandomThreadCard';
import { generateRandomThread, fetchUserSettings, updateUserSettings, fetchActiveThreads, fetchCharacters, updateThread, openUntrackThreadModal } from '../../../infrastructure/actions';
import { getMyTurnThreads, getTheirTurnThreads, getQueuedThreads, getRecentActivity, getThreadCountsByCharacter } from '../../../infrastructure/selectors';

const propTypes = {
	characters: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
	userSettings: PropTypes.shape({
		id: PropTypes.string
	}).isRequired,
	dispatch: PropTypes.func.isRequired,
	myTurnThreads: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
	theirTurnThreads: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
	activeThreads: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
	queuedThreads: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
	recentActivityThreads: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
	randomThread: PropTypes.shape({}).isRequired,
	threadsLoading: PropTypes.bool.isRequired,
	characterThreadCounts: PropTypes.shape({}).isRequired
};

function mapStateToProps(state) {
	const {
		characters,
		userSettings,
		activeThreads,
		randomThread,
		loading
	} = state;
	const myTurnThreads = getMyTurnThreads(state);
	const theirTurnThreads = getTheirTurnThreads(state);
	const queuedThreads = getQueuedThreads(state);
	const recentActivityThreads = getRecentActivity(state);
	const characterThreadCounts = getThreadCountsByCharacter(state);
	return {
		characters,
		userSettings,
		activeThreads,
		randomThread,
		myTurnThreads,
		theirTurnThreads,
		queuedThreads,
		recentActivityThreads,
		characterThreadCounts,
		threadsLoading: loading.threadsLoading
	};
}

class Dashboard extends Component {
	constructor(props) {
		super(props);
		this.showDashboardThreadDistributionToggle =
			this.showDashboardThreadDistributionToggle.bind(this);
		this.generateRandomThread = this.generateRandomThread.bind(this);
		this.archiveThread = this.archiveThread.bind(this);
		this.openUntrackThreadModal = this.openUntrackThreadModal.bind(this);
		this.markThreadQueued = this.markThreadQueued.bind(this);
	}

	componentDidMount() {
		const { dispatch } = this.props;
		if (!this.props.userSettings || !this.props.userSettings.id) {
			dispatch(fetchUserSettings());
		}
		if (!this.props.activeThreads || !this.props.activeThreads.length) {
			dispatch(fetchActiveThreads());
		}
		if (!this.props.characters || !this.props.characters.length) {
			dispatch(fetchCharacters());
		}
	}

	showDashboardThreadDistributionToggle() {
		const { dispatch, userSettings } = this.props;
		dispatch(updateUserSettings({ showDashboardThreadDistribution: !userSettings.showDashboardThreadDistribution }));
	}

	generateRandomThread() {
		const { dispatch } = this.props;
		dispatch(generateRandomThread());
	}

	archiveThread(thread) {
		const { dispatch } = this.props;
		const updatedThread = {
			...thread, isArchived: !thread.isArchived
		};
		dispatch(updateThread(updatedThread));
	}

	markThreadQueued(thread) {
		const { dispatch } = this.props;
		const updatedThread = {
			...thread, dateMarkedQueued: new Date(Date.now())
		};
		dispatch(updateThread(updatedThread));
	}

	openUntrackThreadModal(thread) {
		const { dispatch } = this.props;
		dispatch(openUntrackThreadModal(thread));
	}

	render() {
		const {
			characters,
			userSettings,
			activeThreads,
			myTurnThreads,
			theirTurnThreads,
			queuedThreads,
			recentActivityThreads,
			randomThread,
			threadsLoading,
			characterThreadCounts
		} = this.props;
		return (
			<div className="animated fadeIn dashboard-container">
				<Row>
					<Col>
						<AtAGlanceCard
							showDashboardThreadDistribution={userSettings.showDashboardThreadDistribution}
							showDashboardThreadDistributionToggle={this.showDashboardThreadDistributionToggle}
							myTurnThreads={myTurnThreads}
							theirTurnThreads={theirTurnThreads}
							activeThreads={activeThreads}
							queuedThreads={queuedThreads}
							threadsLoading={threadsLoading}
						/>
					</Col>
				</Row>
				<Row>
					<Col xs="12" md="6">
						<RecentActivityCard threads={recentActivityThreads} archiveThread={this.archiveThread} openUntrackThreadModal={this.openUntrackThreadModal} markThreadQueued={this.markThreadQueued} />
					</Col>
					<Col xs="12" md="6">
						<YourCharactersCard characters={characters} characterThreadCounts={characterThreadCounts} />
					</Col>
				</Row>
				<Row>
					<Col md="6">
						<RandomThreadCard
							generateRandomThread={this.generateRandomThread}
							randomThread={randomThread}
						/>
					</Col>
					<Col md="6">
						<TrackerSupportCard />
					</Col>
				</Row>
			</div >
		);
	}
}

Dashboard.propTypes = propTypes;
export default connect(mapStateToProps)(Dashboard);
