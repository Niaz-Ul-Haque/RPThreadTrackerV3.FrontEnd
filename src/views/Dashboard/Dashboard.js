import React, { Component } from 'react';
import {
	Row, Col, Card, CardGroup, CardHeader, CardBlock, TabContent, TabPane, Nav, NavItem, NavLink,
	Table, Badge, Label, Input
} from "reactstrap";
import classnames from "classnames";
import AtAGlanceCard from '../../components/dashboard/at-a-glance/AtAGlanceCard';
import RecentActivityCard from '../../components/dashboard/recent-activity/RecentActivityCard';

class Dashboard extends Component {

	constructor(props) {
		super(props);

		this.toggle = this.toggle.bind(this);
		this.state = {
			activeTab: '1'
		};
	}

	toggle(tab) {
		if (this.state.activeTab !== tab) {
			this.setState({
				activeTab: tab
			});
		}
	}

	render() {
		return (
			<div className="animated fadeIn dashboard-container">
				<Row>
					<Col>
						<AtAGlanceCard />

					</Col>
				</Row>
				<Row>
					<Col md="6" sm="12" >
						<RecentActivityCard />
					</Col>
					<Col md="6" sm="12">
						<Card className="your-characters-card">
							<CardHeader>
								<i className="fa fa-users"></i> Your Characters
							</CardHeader>
							<CardBlock className="card-body">
								<Table>
									<tbody>
										<tr>
											<td>
												Unnamed Character
											</td>
											<td>
												<a href="#">Edit</a>
											</td>
										</tr>
									</tbody>
								</Table>
							</CardBlock>
						</Card>
					</Col>
				</Row>
				<Row>
					<Col md="6">
						<Card className="random-thread-generator-card">
							<CardHeader>
								<i className="fa fa-random"></i> Random Thread Generator
							</CardHeader>
							<CardBlock className="card-body">
								<p>Pick a random thread to respond to!</p>
								<button className="btn btn-primary">Generate</button>
							</CardBlock>
						</Card>
					</Col>
					<Col md="6">
						<Card className="patreon-card">
							<CardHeader>
								<i className="fa fa-dollar"></i> Support Tracker Development
							</CardHeader>
							<CardBlock className="card-body">
								<div className="form-container">
									<form action="https://www.paypal.com/cgi-bin/webscr" method="post" target="_blank" className="ng-pristine ng-valid">
										<input type="hidden" name="cmd" value="_s-xclick" autoComplete="off" />
										<input type="hidden" name="encrypted" value="-----BEGIN PKCS7-----MIIHNwYJKoZIhvcNAQcEoIIHKDCCByQCAQExggEwMIIBLAIBADCBlDCBjjELMAkGA1UEBhMCVVMxCzAJBgNVBAgTAkNBMRYwFAYDVQQHEw1Nb3VudGFpbiBWaWV3MRQwEgYDVQQKEwtQYXlQYWwgSW5jLjETMBEGA1UECxQKbGl2ZV9jZXJ0czERMA8GA1UEAxQIbGl2ZV9hcGkxHDAaBgkqhkiG9w0BCQEWDXJlQHBheXBhbC5jb20CAQAwDQYJKoZIhvcNAQEBBQAEgYCdkCnTXx/RuVx7HBBL10zqmeJH1XcK0sgNZ6sBSkxf/4f5kLNu83TdGQBvKduPgjIKxu8+yrMgaU0O9JN62h2cqh3Ugb3BcpK6S4zf45unmMqtSlmsG3VOCCU5N75uqL+IywgG5edzifA/kc7HVxXLMlQlQ32QMwtGTQCU2LUNvjELMAkGBSsOAwIaBQAwgbQGCSqGSIb3DQEHATAUBggqhkiG9w0DBwQIb6NQDArCsHyAgZBvsH7rtdCN6OGQ4HP0qqeYcJfkgRUCdGOhhnHxvDK9fB8kwclnh9P04M1Ss+kA7epyOvLayZDv5zojOaUhFbzOY4esT/hsq+CMqQSVxNc5g/fQ3LB/sr7LbAlVgBkj6y/fhpLy3hGXK6NKUsSfiHSLhiusBflpuRJZAvONMBbMnZtOSWQ8jznG5Y/H3+EhCwCgggOHMIIDgzCCAuygAwIBAgIBADANBgkqhkiG9w0BAQUFADCBjjELMAkGA1UEBhMCVVMxCzAJBgNVBAgTAkNBMRYwFAYDVQQHEw1Nb3VudGFpbiBWaWV3MRQwEgYDVQQKEwtQYXlQYWwgSW5jLjETMBEGA1UECxQKbGl2ZV9jZXJ0czERMA8GA1UEAxQIbGl2ZV9hcGkxHDAaBgkqhkiG9w0BCQEWDXJlQHBheXBhbC5jb20wHhcNMDQwMjEzMTAxMzE1WhcNMzUwMjEzMTAxMzE1WjCBjjELMAkGA1UEBhMCVVMxCzAJBgNVBAgTAkNBMRYwFAYDVQQHEw1Nb3VudGFpbiBWaWV3MRQwEgYDVQQKEwtQYXlQYWwgSW5jLjETMBEGA1UECxQKbGl2ZV9jZXJ0czERMA8GA1UEAxQIbGl2ZV9hcGkxHDAaBgkqhkiG9w0BCQEWDXJlQHBheXBhbC5jb20wgZ8wDQYJKoZIhvcNAQEBBQADgY0AMIGJAoGBAMFHTt38RMxLXJyO2SmS+Ndl72T7oKJ4u4uw+6awntALWh03PewmIJuzbALScsTS4sZoS1fKciBGoh11gIfHzylvkdNe/hJl66/RGqrj5rFb08sAABNTzDTiqqNpJeBsYs/c2aiGozptX2RlnBktH+SUNpAajW724Nv2Wvhif6sFAgMBAAGjge4wgeswHQYDVR0OBBYEFJaffLvGbxe9WT9S1wob7BDWZJRrMIG7BgNVHSMEgbMwgbCAFJaffLvGbxe9WT9S1wob7BDWZJRroYGUpIGRMIGOMQswCQYDVQQGEwJVUzELMAkGA1UECBMCQ0ExFjAUBgNVBAcTDU1vdW50YWluIFZpZXcxFDASBgNVBAoTC1BheVBhbCBJbmMuMRMwEQYDVQQLFApsaXZlX2NlcnRzMREwDwYDVQQDFAhsaXZlX2FwaTEcMBoGCSqGSIb3DQEJARYNcmVAcGF5cGFsLmNvbYIBADAMBgNVHRMEBTADAQH/MA0GCSqGSIb3DQEBBQUAA4GBAIFfOlaagFrl71+jq6OKidbWFSE+Q4FqROvdgIONth+8kSK//Y/4ihuE4Ymvzn5ceE3S/iBSQQMjyvb+s2TWbQYDwcp129OPIbD9epdr4tJOUNiSojw7BHwYRiPh58S1xGlFgHFXwrEBb3dgNbMUa+u4qectsMAXpVHnD9wIyfmHMYIBmjCCAZYCAQEwgZQwgY4xCzAJBgNVBAYTAlVTMQswCQYDVQQIEwJDQTEWMBQGA1UEBxMNTW91bnRhaW4gVmlldzEUMBIGA1UEChMLUGF5UGFsIEluYy4xEzARBgNVBAsUCmxpdmVfY2VydHMxETAPBgNVBAMUCGxpdmVfYXBpMRwwGgYJKoZIhvcNAQkBFg1yZUBwYXlwYWwuY29tAgEAMAkGBSsOAwIaBQCgXTAYBgkqhkiG9w0BCQMxCwYJKoZIhvcNAQcBMBwGCSqGSIb3DQEJBTEPFw0xNDAzMTcwMTA5MTdaMCMGCSqGSIb3DQEJBDEWBBRJILNAkTVUgXTmWUFAaKaS/nE5xjANBgkqhkiG9w0BAQEFAASBgHF78RgQC/eZaM1pFdxZ350uTLAK8jxPS9x6k4MA9bfK70EitmPQg0Bq6P8WYs0EHdQ63Emlb2TbjyR+SQmNTGrWdvXXKRUdJqoBjyxSFgY2lzh1vAg9LY1fL9ab2CUoQvhNJTlnJUDFnIJVIlsS9cnG4xtRw81KuzZdYx07yFQ7-----END PKCS7-----" autoComplete="off" />
										<input type="image" src="https://www.paypalobjects.com/en_US/i/btn/btn_donateCC_LG.gif" name="submit" alt="PayPal - The safer, easier way to pay online!" />
										<img alt="" src="https://www.paypalobjects.com/en_US/i/scr/pixel.gif" width="1" height="1" />
									</form>
								</div>
								<a href="https://www.patreon.com/bePatron?u=4797959" className="btn btn-primary">Support me on Patreon!</a>
							</CardBlock>
						</Card>
					</Col>
				</Row>
			</div >
		)
	}
}

export default Dashboard;
