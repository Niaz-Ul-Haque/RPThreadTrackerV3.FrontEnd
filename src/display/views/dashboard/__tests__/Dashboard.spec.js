// #region imports
import React from 'react';
import { getSpecWrapper, shallowWithState, initMockDateNow } from '~/testhelpers/helpers.unit';
import Dashboard from '../Dashboard';
// #endregion imports

// #region mocks
initMockDateNow();
jest.mock('../components/at-a-glance/AtAGlanceCard', () => 'AtAGlanceCard');
jest.mock('../components/recent-activity/RecentActivityCard', () => 'RecentActivityCard');
jest.mock('../components/your-characters/YourCharactersCard', () => 'YourCharactersCard');
jest.mock('../components/tracker-support/TrackerSupportCard', () => 'TrackerSupportCard');
jest.mock('../components/random-thread/RandomThreadCard', () => 'RandomThreadCard');
jest.mock('../../../../infrastructure/actions', () => ({}));
jest.mock('../../../../infrastructure/selectors', () => ({
	getMyTurnThreads: () => [
		{
			threadId: 5,
			userTitle: 'My Turn 1'
		},
		{
			threadId: 6,
			userTitle: 'My Turn 2'
		}
	],
	getTheirTurnThreads: () => [
		{
			threadId: 7,
			userTitle: 'Their Turn 1'
		},
		{
			threadId: 8,
			userTitle: 'Their Turn 2'
		}
	],
	getQueuedThreads: () => [
		{
			threadId: 9,
			userTitle: 'Queued Turn 1'
		},
		{
			threadId: 10,
			userTitle: 'Queued Turn 2'
		}
	],
	getRecentActivity: () => [
		{
			threadId: 11,
			userTitle: 'Recent 1'
		},
		{
			threadId: 12,
			userTitle: 'Recent 2'
		}
	],
	getThreadCountsByCharacter: () => ({
		2: [
			{
				threadId: 13,
				userTitle: 'By Character 1'
			},
			{
				threadId: 14,
				userTitle: 'By Character 2'
			}
		]
	}),
	getIsLoadingIconVisible: () => true
}));
// #endregion mocks

const createTestProps = (propOverrides) => ({
	fetchActiveThreads: jest.fn(),
	fetchCharacters: jest.fn(),
	generateRandomThread: jest.fn(),
	openUntrackThreadModal: jest.fn(),
	updateUserSettings: jest.fn(),
	upsertThread: jest.fn(),
	...propOverrides
});

const createTestState = (stateOverrides) => ({
	characters: [
		{
			characterId: 1,
			characterName: 'Character A'
		},
		{
			characterId: 2,
			characterName: 'Character B'
		}
	],
	userSettings: {
		settingsId: 3
	},
	activeThreads: [
		{
			threadId: 1,
			userTitle: 'Active 1'
		},
		{
			threadId: 2,
			userTitle: 'Active 2'
		},
		{
			threadId: 3,
			userTitle: 'Active 3'
		}
	],
	randomThread: {
		threadId: 2,
		userTitle: 'My Random Thread'
	},
	...stateOverrides
});

describe('rendering', () => {
	describe('snapshots', () => {
		it('should render valid snapshot', () => {
			const props = createTestProps();
			const state = createTestState();
			const jsx = <Dashboard {...props} />;
			const element = shallowWithState(jsx, state).dive();
			expect(element).toMatchSnapshot();
		});
	});
});

describe('behavior', () => {
	describe('componentDidMount', () => {
		it('should fetch characters when characters are not loaded', () => {
			const fetchCharacters = jest.fn();
			const props = createTestProps({
				fetchCharacters
			});
			const state = createTestState({
				characters: []
			});
			const jsx = <Dashboard {...props} />;
			shallowWithState(jsx, state).dive();
			expect(fetchCharacters).toHaveBeenCalledTimes(1);
		});
		it('should not fetch characters when characters are loaded', () => {
			const fetchCharacters = jest.fn();
			const props = createTestProps({
				fetchCharacters
			});
			const state = createTestState();
			const jsx = <Dashboard {...props} />;
			shallowWithState(jsx, state).dive();
			expect(fetchCharacters).toHaveBeenCalledTimes(0);
		});
		it('should fetch active threads when active threads are not loaded', () => {
			const fetchActiveThreads = jest.fn();
			const props = createTestProps({
				fetchActiveThreads
			});
			const state = createTestState({
				activeThreads: []
			});
			const jsx = <Dashboard {...props} />;
			shallowWithState(jsx, state).dive();
			expect(fetchActiveThreads).toHaveBeenCalledTimes(1);
		});
		it('should not fetch active threads when active threads are loaded', () => {
			const fetchActiveThreads = jest.fn();
			const props = createTestProps({
				fetchActiveThreads
			});
			const state = createTestState();
			const jsx = <Dashboard {...props} />;
			shallowWithState(jsx, state).dive();
			expect(fetchActiveThreads).toHaveBeenCalledTimes(0);
		});
	});
	describe('generateRandomThread', () => {
		it('should be triggered when child component prop is triggered', () => {
			const generateRandomThread = jest.fn();
			const props = createTestProps({
				generateRandomThread
			});
			const state = createTestState();
			const jsx = <Dashboard {...props} />;
			const element = shallowWithState(jsx, state).dive();
			const child = getSpecWrapper(element, 'dashboard-random-thread-card');
			child.props().generateRandomThread();
			expect(generateRandomThread).toHaveBeenCalledTimes(1);
		});
	});
	describe('openUntrackThreadModal', () => {
		it('should be triggered when child component prop is triggered', () => {
			const openUntrackThreadModal = jest.fn();
			const props = createTestProps({
				openUntrackThreadModal
			});
			const state = createTestState();
			const jsx = <Dashboard {...props} />;
			const element = shallowWithState(jsx, state).dive();
			const child = getSpecWrapper(element, 'dashboard-recent-activity-card');
			child.props().openUntrackThreadModal();
			expect(openUntrackThreadModal).toHaveBeenCalledTimes(1);
		});
	});
	describe('showDashboardThreadDistributionToggle', () => {
		it('should be triggered when child component prop is triggered', () => {
			const updateUserSettings = jest.fn();
			const props = createTestProps({
				updateUserSettings
			});
			const state = createTestState();
			const jsx = <Dashboard {...props} />;
			const element = shallowWithState(jsx, state).dive();
			const child = getSpecWrapper(element, 'dashboard-at-a-glance-card');
			child.props().showDashboardThreadDistributionToggle();
			expect(updateUserSettings).toHaveBeenCalledTimes(1);
		});
		it('should dispatch user settings update with thread distribution property reversed', () => {
			const updateUserSettings = jest.fn();
			const props = createTestProps({
				updateUserSettings
			});
			const state = createTestState({
				userSettings: {
					settingsId: 5,
					showDashboardThreadDistribution: false
				}
			});
			const jsx = <Dashboard {...props} />;
			const element = shallowWithState(jsx, state).dive();
			element.instance().showDashboardThreadDistributionToggle();
			expect(updateUserSettings).toHaveBeenCalledTimes(1);
			expect(updateUserSettings).toHaveBeenLastCalledWith({
				settingsId: 5,
				showDashboardThreadDistribution: true
			});
		});
	});
	describe('markThreadQueued', () => {
		it('should be triggered when child component prop is triggered', () => {
			const upsertThread = jest.fn();
			const props = createTestProps({
				upsertThread
			});
			const state = createTestState();
			const jsx = <Dashboard {...props} />;
			const element = shallowWithState(jsx, state).dive();
			const child = getSpecWrapper(element, 'dashboard-recent-activity-card');
			child.props().markThreadQueued();
			expect(upsertThread).toHaveBeenCalledTimes(1);
		});
		it('should dispatch thread update with queued date set', () => {
			const upsertThread = jest.fn();
			const props = createTestProps({
				upsertThread
			});
			const state = createTestState();
			const jsx = <Dashboard {...props} />;
			const element = shallowWithState(jsx, state).dive();
			element.instance().markThreadQueued({
				threadId: 1
			});
			expect(upsertThread).toHaveBeenCalledTimes(1);
			expect(upsertThread).toHaveBeenLastCalledWith({
				threadId: 1,
				dateMarkedQueued: new Date(global.Date.now())
			});
		});
	});
	describe('archiveThread', () => {
		it('should be triggered when child component prop is triggered', () => {
			const upsertThread = jest.fn();
			const props = createTestProps({
				upsertThread
			});
			const state = createTestState();
			const jsx = <Dashboard {...props} />;
			const element = shallowWithState(jsx, state).dive();
			const child = getSpecWrapper(element, 'dashboard-recent-activity-card');
			child.props().archiveThread({
				threadId: 1
			});
			expect(upsertThread).toHaveBeenCalledTimes(1);
		});
		it('should dispatch thread update with isArchived reversed', () => {
			const upsertThread = jest.fn();
			const props = createTestProps({
				upsertThread
			});
			const state = createTestState();
			const jsx = <Dashboard {...props} />;
			const element = shallowWithState(jsx, state).dive();
			element.instance().archiveThread({
				threadId: 1,
				isArchived: false
			});
			expect(upsertThread).toHaveBeenCalledTimes(1);
			expect(upsertThread).toHaveBeenLastCalledWith({
				threadId: 1,
				isArchived: true
			});
		});
	});
});
