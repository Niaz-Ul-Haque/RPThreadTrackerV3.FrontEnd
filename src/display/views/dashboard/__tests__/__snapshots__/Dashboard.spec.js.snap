// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering snapshots should render valid snapshot 1`] = `
<div
  className="animated fadeIn dashboard-container"
>
  <Row
    tag="div"
  >
    <Col
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <AtAGlanceCard
        activeThreads={
          Array [
            Object {
              "threadId": 1,
              "userTitle": "Active 1",
            },
            Object {
              "threadId": 2,
              "userTitle": "Active 2",
            },
            Object {
              "threadId": 3,
              "userTitle": "Active 3",
            },
          ]
        }
        data-spec="dashboard-at-a-glance-card"
        isLoadingIconVisible={true}
        myTurnThreads={
          Array [
            Object {
              "threadId": 5,
              "userTitle": "My Turn 1",
            },
            Object {
              "threadId": 6,
              "userTitle": "My Turn 2",
            },
          ]
        }
        queuedThreads={
          Array [
            Object {
              "threadId": 9,
              "userTitle": "Queued Turn 1",
            },
            Object {
              "threadId": 10,
              "userTitle": "Queued Turn 2",
            },
          ]
        }
        showDashboardThreadDistributionToggle={[Function]}
        theirTurnThreads={
          Array [
            Object {
              "threadId": 7,
              "userTitle": "Their Turn 1",
            },
            Object {
              "threadId": 8,
              "userTitle": "Their Turn 2",
            },
          ]
        }
      />
    </Col>
  </Row>
  <Row
    tag="div"
  >
    <Col
      md="6"
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
      xs="12"
    >
      <RecentActivityCard
        allThreads={
          Array [
            Object {
              "threadId": 1,
              "userTitle": "Active 1",
            },
            Object {
              "threadId": 2,
              "userTitle": "Active 2",
            },
            Object {
              "threadId": 3,
              "userTitle": "Active 3",
            },
          ]
        }
        archiveThread={[Function]}
        characters={
          Array [
            Object {
              "characterId": 1,
              "characterName": "Character A",
            },
            Object {
              "characterId": 2,
              "characterName": "Character B",
            },
          ]
        }
        data-spec="dashboard-recent-activity-card"
        loadingInProgress={true}
        markThreadQueued={[Function]}
        openUntrackThreadModal={[MockFunction]}
        recentActivityThreads={
          Array [
            Object {
              "threadId": 11,
              "userTitle": "Recent 1",
            },
            Object {
              "threadId": 12,
              "userTitle": "Recent 2",
            },
          ]
        }
      />
    </Col>
    <Col
      md="6"
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
      xs="12"
    >
      <YourCharactersCard
        characterThreadCounts={
          Object {
            "2": Array [
              Object {
                "threadId": 13,
                "userTitle": "By Character 1",
              },
              Object {
                "threadId": 14,
                "userTitle": "By Character 2",
              },
            ],
          }
        }
        characters={
          Array [
            Object {
              "characterId": 1,
              "characterName": "Character A",
            },
            Object {
              "characterId": 2,
              "characterName": "Character B",
            },
          ]
        }
        loadingInProgress={true}
      />
    </Col>
  </Row>
  <Row
    tag="div"
  >
    <Col
      md="6"
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
      xs="12"
    >
      <RandomThreadCard
        data-spec="dashboard-random-thread-card"
        generateRandomThread={[MockFunction]}
        randomThread={
          Object {
            "threadId": 2,
            "userTitle": "My Random Thread",
          }
        }
      />
    </Col>
    <Col
      md="6"
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
      xs="12"
    >
      <TrackerSupportCard />
    </Col>
  </Row>
</div>
`;
